src/mapreduce.jl

```diff
@@ -0,0 +1,276 @@
+struct Batches{V}
+    firstindex::Int
+    lastindex::Int
+    batch_size::Int
+    values::V
+    length::Int
+end
+
+function Batches(values, batch_size::Integer)
+    @argcheck batch_size > 0
+    r = eachindex(values)
+    @assert r isa AbstractUnitRange
+    len = ceil(Int, length(r) / batch_size)
+    Batches(first(r), last(r), batch_size, values, len)
+end
+
+Base.length(o::Batches) = o.length
+Base.eachindex(o::Batches) = Base.OneTo(length(o))
+function Base.getindex(o::Batches, i)
+    @boundscheck (@argcheck i in eachindex(o))
+    start = o.firstindex + (i-1) * o.batch_size
+    stop  = min(start + (o.batch_size) -1, o.lastindex)
+    o.values[start:stop]
+end
+
+function default_batch_size(len)
+    len <= 1 && return 1
+    nthreads=Threads.nthreads()
+    items_per_thread = len/nthreads
+    items_per_batch = items_per_thread/4
+    clamp(1, round(Int, items_per_batch), len)
+end
+
+function Base.iterate(o::Batches, state=1)
+    if state in eachindex(o)
+        o[state], state+1
+    else
+        nothing
+    end
+end
+
+mutable struct _RollingCutOut{A,I<:AbstractUnitRange,T} <: AbstractVector{T}
+    array::A
+    eachindex::I
+end
+
+function _RollingCutOut(array::AbstractArray, indices)
+    T = eltype(array)
+    A = typeof(array)
+    I = typeof(indices)
+    _RollingCutOut{A, I, T}(array, indices)
+end
+
+Base.size(r::_RollingCutOut) = (length(r.eachindex),)
+
+function Base.eachindex(r::_RollingCutOut, rs::_RollingCutOut...)
+    for r2 in rs
+        @assert r.eachindex == r2.eachindex
+    end
+    Base.IdentityUnitRange(r.eachindex)
+end
+Base.axes(r::_RollingCutOut) = (eachindex(r),)
+
+@inline function Base.getindex(o::_RollingCutOut, i)
+    @boundscheck checkbounds(o, i)
+    @inbounds o.array[i]
+end
+
+@inline function Base.setindex!(o::_RollingCutOut, val, i)
+    @boundscheck checkbounds(o, i)
+    @inbounds o.array[i] = val
+end
+
+############################## tmap, tmap! ##############################
+struct MapWorkspace{F,B,AD,AS}
+    f::F
+    batches::B
+    arena_dst_view::AD
+    arena_src_views::AS
+end
+
+@noinline function run!(o::MapWorkspace)
+    let o=o
+        Threads.@threads for i in 1:length(o.batches)
+            tid = Threads.threadid()
+            dst_view  = o.arena_dst_view[tid]
+            src_views = o.arena_src_views[tid]
+            inds = o.batches[i]
+            dst_view.eachindex = inds
+            for view in src_views
+                view.eachindex = inds
+            end
+            Base.map!(o.f, dst_view, src_views...)
+        end
+    end
+end
+
+function create_arena_src_views(srcs, sample_inds)
+    nt = Threads.nthreads()
+    [Base.map(src -> _RollingCutOut(src, sample_inds), srcs) for _ in 1:nt]
+end
+
+@noinline function prepare(::typeof(tmap!), f, dst, srcs; batch_size::Int)
+    # we use IndexLinear since _RollingCutOut implementation
+    # does not support other indexing well
+    all_inds  = eachindex(IndexLinear(), srcs...)
```

Comment:
What is the advantage of this approach compared to baked in linear indexing which I think we use in `master`?

Reply:
I suppose with this we can handle arrays with `CartesianIndex` indexing style?

Reply:
Yes its with `CartesianIndex` in mind.

Reply:
I have a feeling this may be over-complicating things but I have to sit down and see how to simplify this code. Anyways, I would appreciate a summary of the features that this re-write adds compared to master so I can keep them in mind while reading.

Reply:
If you come up with something simpler that would be great of course. See [here](https://github.com/mohamed82008/KissThreading.jl/issues/22#issue-459398750) for the features. 
Here is why I choose to  implement it like this:
* There is a feature which says e.g. `Base.sum` etc. should be reused.
* This means for `tsum` we need to create subarrays and run `Base.sum` on each
* Specifying a subarray means specifying a bunch of indices. The easiest way of doing so without allocation is to specify ranges, (hence use `IndexLinear`).

Reply:
Ok I see the point now. I will continue the review some time tonight.

Reply:
I wonder if reusing `Base` will be faster than a simple loop or not especially for small batch sizes. I kind of doubt it but please benchmark against master.

Reply:
Yup I also doubt it. In case of sum it is about roundoff errors, as `Base.sum` uses pairwise summation.

Reply:
If speed is significantly hurt, I don't think it is worth the trade-off. We can try to re-implement the pairwise summation instead to get the best of both worlds.

Reply:
Oh I commented before I saw the benchmarks oops!